planner/cascades/transformation_rules.go

```diff
@@ -591,3 +594,52 @@ func (r *PushSelDownJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.Grou
 	newJoinExpr.SetChildren(leftGroup, rightGroup)
 	return []*memo.GroupExpr{newJoinExpr}, true, false, nil
 }
+
+// PushTopNDownProjection pushes TopN to Projection.
+type PushTopNDownProjection struct {
+	baseRule
+}
+
+// NewRulePushTopNDownProjection creates a new Transformation PushTopNDownProjection.
+func NewRulePushTopNDownProjection() Transformation {
+	rule := &PushTopNDownProjection{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandTopN,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandProjection, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// OnTransform implements Transformation interface.
+// This rule tries to pushes the TopN through Projection.
+func (r *PushTopNDownProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	topN := old.GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	proj := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	childGroup := old.Children[0].GetExpr().Children[0]
+
+	newTopN := plannercore.LogicalTopN{
+		Offset: topN.Offset,
+		Count:  topN.Count,
+	}.Init(topN.SCtx(), topN.SelectBlockOffset())
+
+	for _, by := range topN.ByItems {
+		newBy := by.Clone()
```

Comment:
Yes, this `Clone` can be removed. The `ColumnSubstiute` doesn't change the given expr.

Reply:
Can I use the original topN'ByItems to initinal the newTopN? Like:
```
	newTopN := plannercore.LogicalTopN{
		ByItems: topN.ByItems,
		Offset: topN.Offset,
		Count:  topN.Count,
	}.Init(topN.SCtx(), topN.SelectBlockOffset())

	for _, by := range topN.ByItems {
		by.Expr = expression.ColumnSubstitute(by.Expr, old.Children[0].Group.Prop.Schema, proj.Exprs)
	}

```

Reply:
No, the slice should be newed. The old one and the new one can only share the things that is not changed. When subsitute columns, the things in the slice is changed. So we need a new slice.

Reply:
Gota. Thanks

Reply:
You can do it like the following codes:
```
newByItems := alloc len(oldTopN.ByItems)

for _, item := range oldTopN.ByItems {
	newByItems = append(newByItems, &ByItem{
		Expr: ColumnSubtitute()
		Desc: item.Desc
	})
}
```